src/imageio/format/jxl.c

```diff
@@ -0,0 +1,498 @@
+  JXL_ASSERT(JxlEncoderOptionsSetDecodingSpeed(options, decode_effort));
+#endif
+  int encode_effort = params->encoding_effort;
+  // In libjxl 0.5 encode efforts over 4 cause a crash and encode efforts below 3 are unsupported
+  if(JxlEncoderVersion() <= 0 * 1000000 + 5 * 1000 + 0) // if version <= 0.5
+  {
+    fprintf(stderr, "Warning: Encoding effort is limited to 3-4 on libjxl 0.5 and lower\n");
+    if(encode_effort > 4)
+      encode_effort = 4;
+    else if(encode_effort < 3)
+      encode_effort = 3;
+  }
+  JXL_ASSERT(JxlEncoderOptionsSetEffort(options, encode_effort));
+  int quality = params->quality;
+  // We store quality but jxl wants distance, so we reverse the scale
+  quality = (quality - 15) * -1;
+  JXL_ASSERT(JxlEncoderOptionsSetDistance(options, quality));
+
+  int channels = 3;
+  JxlPixelFormat pixel_format = { channels, JXL_TYPE_FLOAT, JXL_NATIVE_ENDIAN, 1 };
+  size_t in_buf_size = width * height * channels * sizeof(float);
+
+  pixels = malloc(in_buf_size);
+  for(int y = 0; y < height; y++)
+  {
+    for(int x = 0; x < width; x++)
+    {
+      const float *buf = &in[((y * width) + x) * 4];
+
+      for(int k = 0; k < channels; k++) pixels[(y * width * channels) + (x * channels) + k] = buf[k];
+    }
+  }
+
+  JXL_ASSERT(JxlEncoderAddImageFrame(options, &pixel_format, pixels, in_buf_size));
+  JxlEncoderCloseInput(encoder);
+
+  // Write the image to disk in chunks of 50,000 bytes
+  // fixme: Can we better estimate what the optimal size of chunks is for this image?
+  size_t chunkSize = 50000;
+
+  out_buf = malloc(chunkSize);
+  if(!out_buf) goto end;
+
+  out_file = g_fopen(filename, "wb");
+  if(!out_file) goto end;
+
+  JxlEncoderStatus out_status = JXL_ENC_NEED_MORE_OUTPUT;
+
+  while(out_status == JXL_ENC_NEED_MORE_OUTPUT)
+  {
+    uint8_t *out_tmp = out_buf;
+    size_t out_avail = chunkSize;
+
+    out_status = JxlEncoderProcessOutput(encoder, &out_tmp, &out_avail);
+    if(out_status != JXL_ENC_NEED_MORE_OUTPUT && out_status != JXL_ENC_SUCCESS)
+    {
+      // out_status should be an error: force it to be processed
+      JXL_ASSERT(out_status);
+      // just to let the compiler know that we will goto end (even though it
+      // should happen anyway in the assertion)
+      goto end;
+    }
+
+    // out_tmp is incremented by the amount written
+    size_t consumed = out_tmp - out_buf;
+    // write however much we got to the file
+    if(fwrite(out_buf, 1, consumed, out_file) != consumed) goto end;
+  }
+
+  // Successful write: set to success code
+  out = 0;
+
+end:
+  if(runner) JxlThreadParallelRunnerDestroy(runner);
+  if(encoder) JxlEncoderDestroy(encoder);
+  if(out_file) fclose(out_file);
+  if(out_buf) free(out_buf);
+  if(pixels) free(pixels);
+
+  return out;
+}
+
+int levels(dt_imageio_module_data_t *data)
+{
+  return IMAGEIO_INT32 | IMAGEIO_FLOAT | IMAGEIO_RGB;
+}
+
+int flags(dt_imageio_module_data_t *data)
+{
+  // fixme: For now exiv2 does not support writing to bmff containers.
```

Comment:
libjxl should support adding the Exif data directly as cjxl seems to do it (see also avif.c which is also not using exiv2 for this)

Reply:
libjxl does not have any API for adding exif data when encoding images, cjxl has custom code to add the exif data.
I will try and do the same.

Reply:
Would be nice to have indeed, as exiv2 probably won't have write support for a while...

Reply:
I have added support for writing exif (as provided by the `exif` parameter to `write_image`) to the outputted file, however XMP is still not supported as the XMP logic happens outside of the image format module so I feel would be out of scope for this PR. As suggested in another comment, however, perhaps in the future I can make a necessary PR to exiv2 to support this. Let me know if this resolves this particular review @kmilos .

Reply:
Yes, this is [the same reason XMP in AVIF is also not supported yet, even though libavif provides and API to write it](https://github.com/darktable-org/darktable/issues/8480)...

I'd love to hear the core devs comment on this PR - it looks like a "complete" and great job to me, but personally I feel a bit uneasy including the extra cruft of that manual BMFF container creation and Exif box writing in dt, there's quite a bit of burden to assure compliance there...

Reply:
> Would be nice to have indeed, as exiv2 probably won't have write support for a while...

The container format is isobmff like avif and heif are using. I think it is just a matter of making sure that exiv2 knows the file extension and then goes to the isobmff parser ...

Reply:
exiv2 cannot write to isobmff containers currently, read only. libheif and libavif have their own methods of writing Exif and XMP boxes. libjxl does not (yet).